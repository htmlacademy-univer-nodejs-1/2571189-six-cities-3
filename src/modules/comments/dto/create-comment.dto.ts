import {IsString, Length, Max, Min} from 'class-validator';

export default class CreateCommentDto {
  @IsString({message: 'text is required'})
  @Length(5, 1024, {message: 'Min length is 5, max is 1024'})
  public text!: string;

  public offerId!: string;

  public userId!: string;

  @Min(1, {message: 'Min value for rating is 1.'})
  @Max(5, {message: 'Max value for rating is 5.'})
  public rating!: number;
}
