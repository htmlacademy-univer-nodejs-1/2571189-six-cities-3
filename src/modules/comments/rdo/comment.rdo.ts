import { Expose, Type } from 'class-transformer';
import UserRdo from '../../user/rdo/user.rdo.js';
import {UserType} from '../../../types/user.type.js';
export default class CommentRdo {
  @Expose()
  public id!: string;

  @Expose()
  public text!: string;

  @Expose()
  public rating!: number;

  @Expose({ name: 'createdAt'})
  public postDate!: string;

  @Expose({name: 'userId'})
  @Type(() => UserRdo)
    user!: UserType;
}
