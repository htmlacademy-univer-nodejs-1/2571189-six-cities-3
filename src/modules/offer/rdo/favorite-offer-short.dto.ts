import { Expose } from 'class-transformer';
import {CityEnum} from '../../../types/city.enum.js';
import {HousingType} from '../../../types/housing-type.enum.js';

export class FavoriteOfferShortDto {
  @Expose()
  public id!: string;

  @Expose()
    name!: string;

  @Expose({ name: 'createdAt'})
    publicationDate!: Date;

  @Expose()
    description!: string;

  @Expose()
    city!: CityEnum;

  @Expose()
    previewImage!: string;

  @Expose()
    premium!: boolean;

  favorite = true;

  @Expose()
    rating!: number;

  @Expose()
    housingType!: HousingType;

  @Expose()
    cost!: number;

  @Expose()
    commentsCount!: number;
}
