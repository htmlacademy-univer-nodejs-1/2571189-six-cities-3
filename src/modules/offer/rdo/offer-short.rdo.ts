import { Expose } from 'class-transformer';
import {CityEnum} from '../../../types/city.enum.js';
import {HousingType} from '../../../types/housing-type.enum.js';

export class OfferShortRdo {
  @Expose()
  public id!: string;

  @Expose()
    name!: string;

  @Expose({ name: 'createdAt'})
    publicationDate!: Date;

  @Expose()
    city!: CityEnum;

  @Expose()
    previewImage!: string;

  @Expose()
    premium!: boolean;

  @Expose()
    favorite!: boolean;

  @Expose()
    rating!: number;

  @Expose()
    housingType!: HousingType;

  @Expose()
    cost!: number;

  @Expose()
    commentsCount!: number;
}
