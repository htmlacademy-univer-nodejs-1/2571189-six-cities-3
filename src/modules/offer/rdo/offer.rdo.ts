import {Expose, Type} from 'class-transformer';
import {CityEnum} from '../../../types/city.enum.js';
import {HousingType} from '../../../types/housing-type.enum.js';
import {Facilities} from '../../../types/facilities.enum.js';
import {UserType} from '../../../types/user.type.js';
import {CoordinatesType} from '../../../types/coordinates.type.js';
import UserRdo from '../../user/rdo/user.rdo.js';

export class OfferRdo {
  @Expose()
  public id!: string;

  @Expose()
    name!: string;

  @Expose({ name: 'createdAt'})
    publicationDate!: Date;

  @Expose()
    description!: string;

  @Expose()
    city!: CityEnum;

  @Expose()
    previewImage!: string;

  @Expose()
    images!: string[];

  @Expose()
    premium!: boolean;

  @Expose()
    favorite!: boolean;

  @Expose()
    rating!: number;

  @Expose()
    housingType!: HousingType;

  @Expose()
    cost!: number;

  @Expose()
    commentsCount!: number;

  @Expose()
    roomCount!: number;

  @Expose()
    guestCount!: number;

  @Expose()
    facilities!: Facilities[];

  @Expose({name: 'userId'})
  @Type(() => UserRdo)
    offerAuthor!: UserType;

  @Expose()
    coordinates!: CoordinatesType;
}
