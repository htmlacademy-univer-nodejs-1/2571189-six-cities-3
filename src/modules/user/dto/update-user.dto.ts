import {IsOptional} from 'class-validator';

export default class UpdateUserDto {
  @IsOptional()
  public email?: string;

  @IsOptional()
  public username?: string;

  @IsOptional()
  public avatar?: string;
}
