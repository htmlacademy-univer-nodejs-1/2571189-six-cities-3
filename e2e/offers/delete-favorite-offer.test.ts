import assert from 'node:assert';
import { Headers, fetch } from 'undici';
import { describe, test } from 'vitest';
import {CityEnum} from '../../src/types/city.enum.js';
import {HousingType} from '../../src/types/housing-type.enum.js';
import {Facilities} from '../../src/types/facilities.enum.js';
import LoggedUserRdo from '../../src/modules/user/rdo/logged-user.rdo.js';
import UserRdo from '../../src/modules/user/rdo/user.rdo.js';
import {OfferRdo} from '../../src/modules/offer/rdo/offer.rdo.js';
import {FavoriteOfferShortDto} from '../../src/modules/offer/rdo/favorite-offer-short.dto';

process.env['E2E_ENDPOINT'] = 'http://localhost:4000';
assert(process.env['E2E_ENDPOINT'] !== undefined);
assert(process.env['E2E_ENDPOINT'].startsWith('http') === true, 'E2E_ENDPOINT должен начинаться с протокола');

const url = new URL(process.env['E2E_ENDPOINT']);

describe('GET /offers/favorites', async () => {
  test('Success get favorite offers after delete', async (tc) => {
    const user = {
      email: 'test1@test.com',
      password: '12345678910',
    };

    const registerResponse = await fetch(new URL('/users/register', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
      }),
      body: JSON.stringify({
        type: 'simple',
        email: 'test1@test.com',
        username: 'name1',
        password: '12345678910',
      })
    });

    const userRdo = await registerResponse.json() as unknown as UserRdo;
    const loginResponse = await fetch(new URL('/users/login', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
      }),
      body: JSON.stringify(user)
    });

    const content: LoggedUserRdo = await loginResponse.json() as unknown as LoggedUserRdo;
    const offer = {
      name: 'nametenlenght',
      description: 'descriptiontwentylenght',
      city: CityEnum.Amsterdam,
      premium: true,
      housingType: HousingType.House,
      roomCount: 2,
      guestCount: 3,
      cost: 10000,
      facilities: [Facilities.AirConditioning],
      userId: '',
      coordinates: {latitude: 2.22222, longitude: 3.444444}
    };

    const createResponse = await fetch(new URL('/offers', url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
        'Authorization': `Bearer ${content.token}`
      }),
      body: JSON.stringify({...offer, userId: userRdo.id})
    });

    tc.expect(createResponse.ok).toBeTruthy();
    const offerRdo = await createResponse.json() as OfferRdo;

    const addToFavoriteResponse = await fetch(new URL(`/offers/favorites/${offerRdo.id}`, url), {
      method: 'POST',
      headers: new Headers({
        'content-type': 'application/json',
        'Authorization': `Bearer ${content.token}`
      })
    });

    tc.expect(addToFavoriteResponse.ok).toBeTruthy();
    tc.expect(addToFavoriteResponse.status).toStrictEqual(204);


    const removeFavoriteResponse = await fetch(new URL(`/offers/favorites/${offerRdo.id}`, url), {
      method: 'DELETE',
      headers: new Headers({
        'content-type': 'application/json',
        'Authorization': `Bearer ${content.token}`
      })
    });
    tc.expect(removeFavoriteResponse.ok).toBeTruthy();
    tc.expect(removeFavoriteResponse.status).toStrictEqual(204);

    const getFavoriteAfterDeleteResponse = await fetch(new URL('/offers/users/favorites', url), {
      method: 'GET',
      headers: new Headers({
        'content-type': 'application/json',
        'Authorization': `Bearer ${content.token}`
      })
    });

    tc.expect(getFavoriteAfterDeleteResponse.ok).toBeTruthy();
    tc.expect(getFavoriteAfterDeleteResponse.status).toStrictEqual(200);
    tc.expect(getFavoriteAfterDeleteResponse.headers.get('content-type')).toMatch(/application\/json/);
    const favoriteOfferRdo = await getFavoriteAfterDeleteResponse.json() as FavoriteOfferShortDto[];
    const res = favoriteOfferRdo.map((x) => ({...x, id: 0}));
    tc.expect(res).toMatchSnapshot();
  });
});
