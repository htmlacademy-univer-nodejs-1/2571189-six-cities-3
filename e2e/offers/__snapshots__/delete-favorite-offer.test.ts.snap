// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GET /offers/favorites > Success get favorite offers after delete 1`] = `[]`;
